Utilize setName as category name instead of generic const Sets

TErenberger opened the pull request:
Before PR: all items used for an itemrack set get put in the same category "Sets". This differs from the adibags_itemrack behavior where each set gets a named category. 

After PR: items found in an itemrack set are grouped in a new category named after the itemrack set

TErenberger committed:
Utilize setName as category name instead of generic const Sets, similar to adibags itemrack

TErenberger committed:
if item is in multiple sets, make new category with combined name

TErenberger commented:
pretty inexperienced with lua/addon dev so feel free to rip it apart. There is some concern about hygiene for the multiple set categories when things are equipped/sets deleted that should probably be addressed before merge. Looking at adibags version, it seemed like categories could be registered with a grouping key, not sure if this is possible with betterbags

TErenberger commented:
Also, perhaps with this method it would be simpler to loop over each item in bag and collect the list of sets they are in rather than the current impl that starts from the itemrack sets.

DJSchaffner commented:
Thank you for your pull request! I was debating whether I should add this functionality, but I guess it is better for the overall visibility. Might be useful to have a toggle to set if you want this functionality or not. However, then the "plugin" characteristic would be lost.

I will use your changes as a baseline to expand on this. Personally, I think using the sets as a starting point is fine. I guess both variants have their downsides. Will have to rewrite how the items are looped and added then though. When going by the inventory slots you probably have to add special triggers for items being picked up and dropped, bank and perhaps other edge cases which could lead to conflicts (Haven't played around with what could be happening here exactly)

DJSchaffner committed:
rename function in preparation for upcoming changes

DJSchaffner committed:
redesigned collection of setnames and added prefix for setnames

DJSchaffner committed:
added todo

DJSchaffner commented:
I redesigned the collection of setnames and added prefixes for the setnames as was the case in the Adibags filter previously.
Did some testing and it seems to work fine for me. Since it's a bit bigger change, would you mind giving it a spin ingame and see if you find some anomalies on your end?

Cheers :)

DJSchaffner committed:
renamed some variables

TErenberger commented:
Sure I can check it out once servers come up.

TErenberger commented:
Items in multiple sets looking good:
![image](https://github.com/DJSchaffner/BetterBags_ItemRack/assets/5725526/5adde272-7f3f-45fc-9326-3da8057d13be)


 however if the bag frame is open when changing sets there is a ghosting effect with the items that were equipped:

![image](https://github.com/DJSchaffner/BetterBags_ItemRack/assets/5725526/b2f51040-691d-4c31-973c-56abcd46c13e)

if the frame is closed and opened again it is resolved.

TErenberger commented:
Another test case: Changing a item that belongs to multiple sets with another. Result currently is that the item that was removed from a set retains the multiple set category even after open/close and equipset/remove set.

Specifically in this SS: the scarlet leggings was part of RET and RET_PVP, but i removed it from RET and replaced with insulated leggings. Scarlet leggings is still part of RET_PVP.

After /reloadui it shows correctly as part of only one set: RET_PVP

![image](https://github.com/DJSchaffner/BetterBags_ItemRack/assets/5725526/b4d4264c-4d91-4e69-b8f2-3017ff147d8e)

DJSchaffner commented:
Thank you for your tests!
I will try to reproduce those later and see if it's fixable on our end here.

> if the frame is closed and opened again it is resolved.

Might have to explicitly trigger a redraw of BetterBags after changing categories. I will explore this and see if that has any sideeffects. Additionally, calling `function categories:DeleteCategory(category)` would trigger a redraw instead of wiping the categories. But that would trigger a redraw multiple times with multiple categories which is probably problematic.

> After /reloadui it shows correctly as part of only one set: RET_PVP

This could very well be a bug on our end here then. Will see where this originates.

DJSchaffner commented:
Another side effect of these changes is that the previous uniformal `Sets` category will still exist after an update is deployed. So hardcoding a deletion of that category is undesirable but seems like it might be the only option right now.
In the future, BetterBags is planning to add more options for plugin categories like temporary categories, which will only exist during a single session. This could also help with old gearsets which are no longer existing, for which the category will stay though.

An alternative solution would be a persistent storage of created categories.

DJSchaffner committed:
made printchat more robust

DJSchaffner committed:
fixed call to WipeCategories and added logging to it

DJSchaffner committed:
now properly use table as a set (prevents unnecessary calls for wipe)

DJSchaffner committed:
changed some logging

DJSchaffner commented:
Looks like I messed something up while refactoring a variable at the end (classic..)
I was however, not able to reproduce the ghosting bug you found with some of your items. Not sure what that's about. Also, couldn't really explain it from the code alone.

A sidenote which seems to be caused by a new BetterBags patch is, that now when changing the category of an item, it will sometimes not be updated in the bag until ordered to redraw. (closing it sometimes doesn't help either) So you actually need to right-click the bag to sort it, or do a reload. Then the changes will apply. I'm asking the BB dev for info on this, since I'm not sure it would be a good idea to trigger redraws within the plugin code.

Example to reproduce:
Item `A` is attached to gear set `GA` and `GB` -> remove `A` from `GA` -> Item will still remain in category `GA + GB` until sorting the bag or something else happens.

DJSchaffner committed:
bump version

DJSchaffner commented:
After some more testing, things are looking fine. So I would merge this request to not blow it up with too many new things. It does what it was supposed to. I will add 1 or 2 more things and probably prepare a release tonight.

If you find more issues until then feel free to post here or open a new issue and I will look at it.

Thanks again for contributing 😄

TErenberger commented:
Awesome, thanks!
